### PR TITLE
test(node): Cover textual repository nodes

### PR DESCRIPTION
**Pull request by fviernau**
See individual commits.

### CODE REVIEW

File: plugins/package-managers/node/src/test/kotlin/utils/NpmSupportTest.kt [hunk 1]
```diff
@@ -196,7 +195,7 @@ class NpmSupportTest : WordSpec({
             @Suppress("Wrapping")
             val node = ObjectMapper().run {
                 createObjectNode().apply {
-                    replace("gitHead", TextNode(HashAlgorithm.SHA1GIT.emptyValue))
+                    replace("gitHead", TextNode("bar"))
```
Comment (sschuberth) [inline plugins/package-managers/node/src/test/kotlin/utils/NpmSupportTest.kt]: Commit message: This is not about the "VCS path", but the "VCS revision".

File: model/src/main/kotlin/Mappers.kt [hunk 2]
```diff
@@ -73,3 +73,5 @@ val EMPTY_JSON_NODE: JsonNode = MissingNode.getInstance()
 inline fun <reified T> String.fromYaml(): T = yamlMapper.readValue(this)
 
 fun Any?.toYaml(): String = yamlMapper.writeValueAsString(this)
+
+fun String.readJsonTree(): JsonNode = jsonMapper.readTree(this)
```
Comment (sschuberth) [inline model/src/main/kotlin/Mappers.kt:77]: Commit message: s/programatical/programatic/

File: plugins/package-managers/node/src/test/kotlin/utils/NpmSupportTest.kt [hunk 3]
```diff
@@ -192,17 +193,16 @@ class NpmSupportTest : WordSpec({
...
-            @Suppress("Wrapping")
-            val node = ObjectMapper().run {
-                createObjectNode().apply {
-                    replace("gitHead", TextNode("bar"))
-                    replace("repository", createObjectNode().apply {
-                        replace("type", TextNode("Git"))
-                        replace("url", TextNode("https://example.com/"))
-                        replace("directory", TextNode("foo"))
-                    })
+            val node = """
```
Comment (sschuberth) [inline plugins/package-managers/node/src/test/kotlin/utils/NpmSupportTest.kt:196]: It'd be some work, but I guess ideally this should be done for all tests in this file now, or? That would nice decouple the test from Jackson as well.

File: plugins/package-managers/node/src/main/kotlin/utils/NpmSupport.kt [hunk 4]
```diff
@@ -153,16 +153,25 @@ internal fun parseNpmVcsInfo(node: JsonNode): VcsInfo {
...
+    return if (repository.isTextual) {
+        VcsInfo(
+            type = VcsType.UNKNOWN,
+            url = expandNpmShortcutUrl(repository.textValue()),
+            path = "",
+            revision = head
+        )
+    } else {
+        val type = repository["type"].textValueOrEmpty()
+        val url = repository.textValue() ?: repository["url"].textValueOrEmpty()
```
Comment (sschuberth) [inline plugins/package-managers/node/src/main/kotlin/utils/NpmSupport.kt:165]: I believe the original `repository.textValue()` here already handles the `repository.isTextual` case, or?

At least I'm pretty sure that original code already intended to fully handle that case:
- `type` would be empty, so `VcsType.forName(type)` returns `VcsType.UNKNOWN`
- `path` would be empty as well

So what additional case does your new code cover?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: plugins/package-managers/node/src/main/kotlin/utils/NpmSupport.kt (hunk 4 above)
Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/utils/NpmSupport.kt:165]: HM, are you sure that the previous code already worked in this case?

When I wrote this code (in my other PR) I do re-call that the analyzer previously did not detect VCS info but afterwards did. So, thought it was needed but probably I made a mistake during testing and had made further changes which made it work.

Comment (fviernau) [reply plugins/package-managers/node/src/main/kotlin/utils/NpmSupport.kt]: Test proofs that you were right.

File: plugins/package-managers/node/src/test/kotlin/utils/NpmSupportTest.kt (hunk 3 above)
Comment (fviernau) [reply plugins/package-managers/node/src/test/kotlin/utils/NpmSupportTest.kt:196]: I would do a follow up PR for this

**Review by sschuberth — APPROVED** [review]

File: plugins/package-managers/node/src/test/kotlin/utils/NpmSupportTest.kt (hunk 3 above)
Comment (fviernau) [reply plugins/package-managers/node/src/test/kotlin/utils/NpmSupportTest.kt:196]: here you go: https://github.com/oss-review-toolkit/ort/pull/8518

**Pull request merged**